### ISSUE TITLE
Add support for pre-release labels

### ISSUE THREAD

**Issue by eduherminio** [skarpdev/dotnet-version-cli#33]
When setting a specific version, it would be awesome to be able to:
* Set a semantic version with [pre-release labels](https://semver.org/#summary) (`-alpha`, etc.).
* Override an existing version that has a pre-release label.

For `major`, `minor` and `patch` behavior could remain the same as it is, not being able to run them successfully if the existing version has a pre-release label.

**Comment by nover** [comment]
@eduherminio 
I like the overall concept of your suggestion, however 
> Override an existing version that has a pre-release label.

I'm not sure whether I like that - to me, `git tags` (which the tool produces as well) are immutable, so replacing a tag is a no go in my book. We could however make it an opt-in / warn the user appropriately before allowing it.

Another approach we could take is looking at how `yarn` / `npm` is doing it given that this tool takes inspiration from them:
https://classic.yarnpkg.com/en/docs/cli/version/#toc-commands

Basically there's `premajor`, `preminor`, `prepatch` and then the `prerelease` toggle to just "increment the pre-release version".

Let me know what you think.

**Comment by eduherminio** [comment]
Hi @nover.

I didn't express myself correctly. By `override` I meant `be able to replace`.
That is, it you currently set a project version to `0.1.0-alpha`, the tool is unable to modify it.
If you add pre-release labels support, it has to be added to both the version that is going to be applied and the existing version.

Messing around with git tags is definitely a no-go 😆 .

Also, I like the idea of `premajor`, `preminor` and `prepatch` with the `<pre-identifier>` as a complement to the labels support in specific version, that would complete my suggested feature

**Comment by nover** [comment]
Well that clarifies it for me - great that we are in the same boat! :D

I will proceed with the `yarn` inspired way of handling pre-releases and "snapping back out of them" - which is correct - that is not handled by the CLI at this point.

I have a quite busy work-schedule these days, but I will do what I can to squeeze it in.

**Comment by eduherminio** [comment]
If you can release the already implemented functionality, I'd be happy to beta-test it.

**Comment by nover** [comment]
Thanks, I seem to remember that I had some problem with the code that I wanted to debug.

I'll get back to you later today!

**Comment by nover** [comment]
@eduherminio pre-major `2.0.0-0` is now rolling (made with the cli tool!)

**Comment by nover** [comment]
Scratch that - there is some problem with publishing the package to nuget.

**Comment by nover** [comment]
@eduherminio Managed to get a version out to nuget now! `v2.0.0-3`, you should be able to install it by

```
$ dotnet tool uninstall -g dotnet-version-cli  
$ dotnet tool install -g dotnet-version-cli --version "2.0.0-*"                                                                                                                                               
```

I'm hitting an interesting nightmare with nuget and semver 2.0.
If I release pacakges like `2.0.0-3` it does not believe it's semver 2.0 so it validates against semver 1.0 rules, and becomes invalid .

So for this particular release I included `build-meta` to force nuget to recognize the version number as semver 2.0.
I will be looking at workarounds for this - so the release of `dotnet version cli` boiled down to the command:
```
$ dotnet version -f src/dotnet-version.csproj --build-meta `git rev-parse --short HEAD` prerelease 
```
To bump the prerelease counter. To enter pre-release mode you can use one of the `prepatch, preminor, premajor` switches :)

**Comment by eduherminio** [comment]
Really appreciate it, I'll test it with my pipelines and get back to you sooner than later.

Regarding your small nightmare, I'd guess it's caused by your pre-release label starting with a number. I've never experienced such problems with tags like [alpha-n](https://www.nuget.org/packages/SheepTools/0.1.0-alpha2), it may be worth a try to ensure it's not that.

**Comment by nover** [comment]
> Really appreciate it, I'll test it with my pipelines and get back to you sooner than later.
Great thanks!

> .. pre-release label starting with a number...
Yes it's definitely that - however `2.0.0-1` _is_ valid SemVer 2.0 (and how yarn does it).  
According to NuGet docs I either have to include `build-meta` or have a `semver 2` dependency version constraint in the nuspec file, so I'm leaning towards the latter.

**Comment by eduherminio** [comment]
True, I understand that when using `prepatch`, `preminor`, or `premajor` you either need something like that, creating a convention that makes someone unhappy or somehow allowing the user to choose a prefix.

Having said that, I've never seen pre-release labels with just numbers in NuGet packages before, therefore I find them confusing and plan to avoid them in my own packages at all costs (my current use case for your tool implies providing the tag anyway); but that's just a point of view I guess 😄

**Comment by nover** [comment]
Yeah I get your point of view, and I am open to improving on this. I agree with not having seen versions like that on nuget, but that is probably due to the constraints imposed by nuget.

I think some of my work-in-progress code has a `prefix` option but it got removed again.
It could be a viable option, which would result in some additional work on the `prerelease`  command to allow the automatic increments.

**Comment by nover** [comment]
@eduherminio 
CLI now defaults to creating pre-releases with a `next` prefix, which can be overriden by supplying the `--prefix` flag. Subsequent `prerelease` invokes should respect whatever prefix is already set.

New version you can test with: `v2.0.0-next.5`.

**Comment by eduherminio** [comment]
I'm installing the latest version, but internally it looks like it's v1.1.2 (and behaves like that). I observed the same behavior in both Windows and Linux.

\> dotnet tool install -g dotnet-version-cli --version "2.0.0-*"
```
You can invoke the tool using the following command: dotnet-version
Tool 'dotnet-version-cli' (version '2.0.0-next.5') was successfully installed.
```
\> dotnet version -f .\\FileParser.csproj --version 1.6.0-alpha.1
```
dotnet-version-cli version 1.1.2.0
ERR Invalid SemVer version string: --version (Parameter 'versionString')
...
```
Checking the binary version in `%USERPROFILE%\.dotnet\tools` (Windows) it also says v1.1.2.

Am I missing something here?

**Comment by eduherminio** [comment]
But the most important bit is that there's a misbehavior:
* When I try to bump a version of an existing package that has a `Version` tag, `v2.0.0-next.5` doesn't replace it, but creates a new `PackageVersion` one (see [this commit](https://github.com/eduherminio/FileParser/commit/6023a906c8ddeffb78767b6f41506dabcf1bec6f), manually done after running the command in its message). 

Also, new version formats `.csproj` file, which didn't happen before and I'd say it's not desirable (can also be observed in that commit).

**Comment by nover** [comment]
Thanks for testing - we're getting there :)

> I've installed the latest version, but internally it looks like it's v1.1.2

Yes that is due to the `PackageVersion` and `Version` field. Long story short, it used to be so that `Version` could not contain a SemVer. Seems like that has changed so I updated to use just the `Version` field.

`v2.0.0-next.6` is on the way - you should remove the `PackageVersion` manually from your affected code, and then please try again.

**Comment by nover** [comment]
> Also, new version formats .csproj file, which didn't happen before and I'd say it's not desirable (can also be observed in that commit).

Ah, yes that is an side-effect of the changes that it's now keeping the csproj as an XDocument in memory to make it's changes, which is then flushed out. Version 1.x did some simple text search and replace, however that caused the problem described in #28

**Comment by eduherminio** [comment]
Cool, `PackageVersion` issue is solved and this version looks good to me overall, testing it locally. I may test it more extensively in my pipelines in the weekend.

**Comment by eduherminio** [comment]
`2.0.0-next.7` 🚀 🚀

**Comment by nover** [comment]
> 2.0.0-next.7 🚀 🚀

The testing over the weekend has been successful @eduherminio  ?  :)

**Comment by eduherminio** [comment]
Yes, look good to me for releasing, I opened #46 but rather tackle (or discard it) separately.
